design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+* Terminate push query (via the `/close-query` endpoint)
+
+The purpose of this KLIP is to reach agreement on the interfaces / public APIs.
+Implementation details will not be covered.
+
+## What is not in scope
+
+This KLIP does not cover Java client support for the following:
+* `DESCRIBE <stream/table>`, `DESCRIBE CONNECTOR`, `DESCRIBE FUNCTION`
+* `EXPLAIN <queryId>`
+* `PRINT <topic>`
+* `SHOW TOPICS EXTENDED`, `SHOW <STREAMS/TABLES> EXTENDED`, `SHOW QUERIES EXTENDED`
+* `CREATE TYPE`, `DROP TYPE`, `SHOW TYPES`
+* `SHOW FUNCTIONS`, `SHOW PROPERTIES`
+* `RUN SCRIPT`
+* Use of other endpoints (info, healthcheck, terminate cluster, status, etc.)
+
+We can always add support for these operations in the future if desired.
+
+As above, implementation details are out of scope as the purpose of this KLIP is to reach agreement
+on the interfaces / public APIs.
+
+## Value/Return
+
+See [KLIP 15](./klip-15-new-api-and-client.md).
+
+## Public APIS
+
+The following subsections describe the methods of the `Client` interface:
+```
+public interface Client {
+    ...
+}
+```
+
+### Constructors
+```
+  static Client create(ClientOptions clientOptions) {
+    return new ClientImpl(clientOptions);
+  }
+
+  static Client create(ClientOptions clientOptions, Vertx vertx) {
+    return new ClientImpl(clientOptions, vertx);
+  }
+```
+
+The Java client will be implemented as a Vert.x HttpClient. We expose a constructor that allows users to provide their own `Vertx` instance
+in order to take advantage of a shared connection pool and other properties if desired.
+
+`ClientOptions` will initially be as follows:
+```
+public interface ClientOptions {
+
+  ClientOptions setHost(String host);
+
+  ClientOptions setPort(int port);
+
+  ClientOptions setUseTls(boolean useTls);
+
+  ClientOptions setUseClientAuth(boolean useClientAuth);
+
+  ClientOptions setTrustStore(String trustStorePath);
+
+  ClientOptions setTrustStorePassword(String trustStorePassword);
+
+  ClientOptions setKeyStore(String keyStorePath);
+
+  ClientOptions setKeyStorePassword(String keyStorePassword);
+
+  ClientOptions setBasicAuthCredentials(String username, String password);
+
+  String getHost();
+
+  int getPort();
+
+  boolean isUseTls();
+
+  boolean isUseClientAuth();
+
+  boolean isUseBasicAuth();
+
+  String getTrustStore();
+
+  String getTrustStorePassword();
+
+  String getKeyStore();
+
+  String getKeyStorePassword();
+
+  String getBasicAuthUsername();
+
+  String getBasicAuthPassword();
+
+  ClientOptions copy();
+
+  static ClientOptions create() {
+    return new ClientOptionsImpl();
+  }
+}
+```
+
+We can always add additional configuration options later. We may also wish to expose the Vert.x `HttpClientOptions` for
+advanced users that wish to provide custom configs.
+
+### Transient queries -- Streaming
+
+The `Client` interface will provide the following methods for streaming the results of a transient (push or pull) query:
```

Comment:
What do you mean by "streaming the result of a pull query" ? Only push queries return a data stream?

Having said that, would it make sense to have two different methods for both query types, as both should return a different "object"?

Reply:
"Streaming the result" just means receiving records one at a time. Today pull queries only return at most one record so the difference between streaming the result and not isn't meaningful, but this could change in the future.

Reply:
Iteration through a result set of a pull query, is _not_ streaming IMHO. Otherwise, each time you use an `Iterator` you would stream data.

Reply:
Whether a result is streamed or obtained as a batch is orthogonal as to whether it's a pull or a push query. You could have a pull query that returns millions of rows so it's not feasible to load them all in memory at once - you just want to spool them to disk. You could have a push query where you just want to receive the next 10 changes that happen to a table (push query with limit), so you want to receive the results as a batch.

Reply:
I did not say anything about the internal implementation. I totally agree that buffering the full result for a large result of a pull query client side would not be a good implementation.

From a user point of view, there is `poll()` or `subscribe()` and internally there are two "transport mechanism we can pick -- and we can implement all 4 (at least 3; not sure if "batch transfer" makes sense for `subscribe()`), and the user should not care.

I am just wondering if we should offer an API that mixes `pull()` and `subscribe()` and allow both for push and pull queries. IMHO, it convolutes the API. I guess it makes sense to allow people to `poll()` or `subscribe()` to a stream (ie, push query -- even if using `poll()` on a _push_ shows that the name "push query" is really terrible... anyway). However, subscribing to a pull query does not really make sense to me because there is no data stream (even if the large data set might be transmitted incrementally from the server to the client).

Reply:
I think you're getting hung up / confused about the word "stream" here. Whether something is "streamed" to a user or not, is nothing to do with the definition of "stream" in the ksqlDB mental model. They're different things.

Reply:
Well. Because we have data streams, maybe the wording should be done more carefully to avoid confusion?